src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,505 @@
+// Copies the name of the given glyph, or, if no name is available, a string of the form gXXXX into the given string.
+fn font_glyph_to_string(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    glyph: u32,
+    str: u32,
+    len: u32,
+) {
+    // len is apparently the assigned heap memory. It seems I should not allocate more than that.
+    // Should not assume I am not writing over anything.
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let name = caller
+        .data()
+        .font
+        .glyph_name(GlyphId(glyph as u16))
+        .map(ToOwned::to_owned)
+        .unwrap_or(format!("g{glyph:4}"));
+
+    let name = CString::new(name).unwrap();
+    let name = &name.as_bytes()[..len as usize]; // are names ever non ascii?
+
+    memory.write(caller.as_context_mut(), str as usize, name);
+}
+
+// fn font_get_glyph_h_advance(font: u32, glyph: u32) -> i32;
+// Returns the default horizontal advance for the given glyph ID the current scale and variations settings.
+fn font_get_glyph_h_advance(caller: Caller<'_, ShapingData>, _font: u32, glyph: u32) -> i32 {
+    caller.data().font.glyph_h_advance(GlyphId(glyph as u16))
+}
+
+// fn font_get_glyph_v_advance(font: u32, glyph: u32) -> i32;
+// Returns the default vertical advance for the given glyph ID the current scale and variations settings.
+fn font_get_glyph_v_advance(caller: Caller<'_, ShapingData>, _font: u32, glyph: u32) -> i32 {
+    caller.data().font.glyph_v_advance(GlyphId(glyph as u16))
+}
+
+// fn font_copy_glyph_outline(font: u32, glyph: u32, outline: *mut CGlyphOutline) -> bool;
+// Copies the outline of the given glyph ID, at current scale and variation settings, into the outline structure provided.
+fn font_copy_glyph_outline(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    glyph: u32,
+    outline: u32,
+) -> u32 {
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let builder = CGlyphOutline {
+        n_points: todo!(),
+        points: todo!(),
+        n_contours: todo!(),
+        contours: todo!(),
+    };
+    // also no clue what to do here
+    // let my_ol = caller.data().outline_glyph(GlyphId(glyph as u16), builder); // ??
+
+    let Ok(()) = memory.write(caller, outline as usize, todo!("builder result goes here")) else {
+        return 0;
+    };
+
+    1
+}
+
+// fn face_copy_table(font: u32, tag: u32, blob: *mut Blob) -> bool;
+// Copies the binary data in the OpenType table referenced by tag into the supplied blob structure.
+fn face_copy_table(mut caller: Caller<'_, ShapingData>, _font: u32, tag: u32, blob: u32) -> u32 {
+    // So here to copy stuff INTO the module, I need to copy it into its heap
+    // I should not assume that there is an area that's not written to,
+    // so the most straightforward way to get "clean" memory is to grow it by one page,
+    // and allocate there.
```

Comment:
yeah, it's very confusing how this is supposed to work.

maybe @behdad can shed some light on this? I see that harfbuzz uses some `wasm_runtime_addr_app_to_native` construct from `wasm-micro-runtime`, but I don't understand how it works conceptually.

Reply:
[This is the function from `wasm-micro-runtime](https://github.com/bytecodealliance/wasm-micro-runtime/blob/0feae05379feb856f6b1af6c0dc5cf57f87155b7/core/iwasm/common/wasm_memory.c#L448).

Reply:
Okay, so that function simply translates the pointer by the base offset. The interesting one that is used to actually write is `wasm_runtime_module_dup_data`, which ends up calling [`wasm_module_malloc_internal`](https://github.com/bytecodealliance/wasm-micro-runtime/blob/0feae05379feb856f6b1af6c0dc5cf57f87155b7/core/iwasm/interpreter/wasm_runtime.c#L3520), which seems to, depending on the circumstances, use various strategies to allocate some memory it can use.